pennylane/pauli/dla/lie_closure.py

```diff
@@ -0,0 +1,257 @@
+        [
+            [1, 1, 1],
+            [1, 0, 0],
+            [0, 1, 0],
+            [0, 0, 0.5]
+        ]
+
+    where each column represents one sentence, and each row represents the coefficient of the respective word in the sentence.
+    To make sense of this representation one additionally needs to keep track of the mapping between keys and rows. In this case we have
+
+    .. code-block::python3
+        pw_to_idx = {
+            X(0) : 0,
+            X(1) : 1,
+            X(2) : 2,
+            Y(0) : 3
+        }
+
+    where we have set the numbering based on appearance in the list of generators. This mapping is in general not unique.
+
+    Args:
+        generators [Iterable[Union[PauliWord, PauliSentence, Operator]]]: Operators that span the vector space.
+
+    **Example**
+
+    Take a linearly dependent set of operators and span the PauliVSpace.
+
+    .. code-block::python3
+        ops = [
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+                PauliWord({0:"Y", 1:"Y"}) : 1.
+            }),
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+            }),
+            PauliSentence({
+                PauliWord({0:"Y", 1:"Y"}) : 1.,
+            }),
+        ]
+
+        vspace = PauliVSpace(ops)
+
+    It automatically detects that the third operator is linearly dependent on the former two, so it does not add the third operator to the basis.
+
+    >>> vspace.basis
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1)]
+
+    We can also retrospectively add operators.
+
+    >>> vspace.add(PauliWord({0:"X"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+
+    Again, checks of linear independence are always performed. So in the following example no operator is added.
+
+    >>> vspace.add(PauliWord({0:"Y", 1:"Y"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+    """
+
+    def __init__(self, generators):
+
+        # Get all Pauli words that are present in at least one Pauli sentence
+        all_pws = list(reduce(set.__or__, [set(ps.keys()) for ps in generators]))
+        num_pw = len(all_pws)
+        # Create a dictionary mapping from PauliWord to row index
+        self._pw_to_idx = {pw: i for i, pw in enumerate(all_pws)}
+
+        # Initialize PauliVSpace properties trivially
+        self._basis = []
+        rank = 0
+
+        self._M = np.zeros((num_pw, rank), dtype=float)
+        self._rank = rank
+        self._num_pw = num_pw
+
+        # Add all generators that are linearly independent
+        self.add(generators)
+
+    @property
+    def basis(self):
+        """List of basis operators of PauliVSpace"""
+        return self._basis
+
+    def __len__(self):
+        return len(self.basis)
+
+    def add(self, other):
+        """Adding a list of PauliSentences if they are linearly independent"""
```

Comment:
This could do with a similar level of detail as `is_independent`

Reply:
✅